### PR TITLE
Improve update method to avoid permanent diff

### PR DESCRIPTION
**Pull request by ziyue-101**
### Change description

Given that the server defaults values if certain fields are not set, the PR  Improves update method to avoid permanent diff

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

**Comment by acpana** [comment]
/lgtm
/approve

thanks for adding these!

cc @maqiuyujoyce would you mind approving too 🙏🏼 ?

The spelled out comparison is necessary since these protos don't implement the message interfaces which would allow us to use the common tooling that yuwen wrote. Note that since there's no changes to http logs this means these don't cause any extra diffs.

Once we merge this Ziyue will look at #2657

File: pkg/controller/direct/gkehub/featuremembership_controller.go [hunk 1]
```diff
@@ -249,7 +248,7 @@ func (a *gkeHubAdapter) Update(ctx context.Context, updateOp *directbase.UpdateO
 	log.V(2).Info("updating object", "u", u)
 	actual := a.actual.MembershipSpecs[a.membershipID]
 	//  There are no output fields in the api Object, so we can compare the desired and the actaul directly.
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/gkehub/featuremembership_controller.go:250]: Just curious, is this comment still valid?

File: pkg/controller/direct/gkehub/featuremembership_controller.go [hunk 2]
```diff
@@ -249,7 +248,7 @@ func (a *gkeHubAdapter) Update(ctx context.Context, updateOp *directbase.UpdateO
...
-	if !reflect.DeepEqual(a.desired.Configmanagement, actual.Configmanagement) || !reflect.DeepEqual(a.desired.Policycontroller, actual.Policycontroller) || !reflect.DeepEqual(a.desired.Mesh, actual.Mesh) {
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/gkehub/featuremembership_controller.go:252]: What's the problem of DeepEqual here? I think we use it a lot in diffing for other resources. If there are any particular issues, we should identify them here and avoid them for future implementations.

**Review by maqiuyujoyce — COMMENTED** [review]
/approve

I have a couple of comments and both comments can be addressed offline.

File: pkg/controller/direct/gkehub/featuremembership_controller.go (hunk 2 above)
Comment (ziyue-101) [reply pkg/controller/direct/gkehub/featuremembership_controller.go:252]: the `desired `is converted from KRM representation. The `actual` is actually fetched from the gcp server. If there are server defaults values, there will be permanent diff in the controller.

File: pkg/controller/direct/gkehub/featuremembership_controller.go (hunk 1 above)
Comment (ziyue-101) [reply pkg/controller/direct/gkehub/featuremembership_controller.go:250]: It still holds in some sense. but we need to rephrase it a bit, will send out a separate PR

**Comment by maqiuyujoyce** [comment]
@ziyue-101 Looking at the diffing functions again, it seems that the diffing works if we support externally-managed fields (i.e. desired state is nil and actual state is non-nil). But in direct controller, I think we want the desired state to be the source of truth, so if the desired state is nil, the actual state should be validated as well: it should be nil or the default value (depending on the API behavior). @yuwenma just want to double check, is my understanding correct?

**Comment by ziyue-101** [comment]
While waiting on yuwen's response, I am wondering if KCC doesn't want to support externally-managed fields, how should the controller handle server populated fields such as defaults? A way could be to record a copy of fields that have defaults, but that would make it harder to keep that copy up-to-date.

It would also be great to update https://cloud.google.com/config-connector/docs/concepts/managing-fields-externally to reflect the latest KCC status.

**Pull request merged**